Sources/YUL/AST.swift

```diff
@@ -0,0 +1,342 @@
+
+    var ret = ""
+    if !self.returns.isEmpty {
+      let retargs = render(typedIdentifierList: self.returns)
+      ret = "-> \(retargs)"
+    }
+
+    return "\(self.identifier)(\(args)) \(ret) \(self.body)"
+  }
+}
+
+public struct VariableDeclaration: CustomStringConvertible {
+  public let declarations: TypedIdentifierList
+  public let expression: Expression?
+
+  public init(_ declarations: TypedIdentifierList, _ expression: Expression? = nil) {
+    self.declarations = declarations
+    self.expression = expression
+  }
+
+  public var catchableSuccesses: [Expression] {
+    return expression?.catchableSuccesses ?? []
+  }
+
+  public var description: String {
+    let decls = render(typedIdentifierList: self.declarations)
+    if self.expression == nil {
+      return "let \(decls)"
+    }
+    return "let \(decls) := \(self.expression!.description)"
+  }
+}
+
+public typealias IdentifierList = [Identifier]
+
+public struct Assignment: CustomStringConvertible {
+  public let identifiers: IdentifierList
+  public let expression: Expression
+
+  public init(_ identifiers: IdentifierList, _ rhs: Expression) {
+    self.identifiers = identifiers
+    self.expression = rhs
+  }
+
+  public var catchableSuccesses: [Expression] {
+    return expression.catchableSuccesses
+  }
+
+  public var description: String {
+    let lhs = self.identifiers.joined(separator: ", ")
+    return "\(lhs) := \(self.expression)"
+  }
+}
+
+public indirect enum Expression: CustomStringConvertible {
+  case functionCall(FunctionCall)
+  case identifier(Identifier)
+  case literal(Literal)
+  case catchable(value: Expression, success: Expression)
+
+  // TODO: these three should really be statements
+  case variableDeclaration(VariableDeclaration)
+  case assignment(Assignment)
+  case noop
+
+  case inline(String)
+
+  public var catchableSuccesses: [Expression] {
+    switch self {
+    case .variableDeclaration(let decl):
+      return decl.catchableSuccesses
+    case .assignment(let assign):
+      return assign.catchableSuccesses
+    case .functionCall(let f):
+      return f.catchableSuccesses
+    case .catchable(_, let success):
+      return [success]
+    default:
+      return []
+    }
+  }
+
+  public var description: String {
+    switch self {
+    case .functionCall(let call):
+      return call.description
+    case .identifier(let id):
+      return id
+    case .literal(let l):
+      return l.description
+    case .variableDeclaration(let decl):
+      return decl.description
+    case .assignment(let assign):
+      return assign.description
+    case .catchable(let value, _):
+      return value.description
+    case .noop:
+      return ""
+    case .inline(let s):
+      return s
+    }
+  }
+}
+
+public struct FunctionCall: CustomStringConvertible {
+  public let name: Identifier
+  public let arguments: [Expression]
+
+  public init(_ name: Identifier, _ arguments: [Expression]) {
```

Comment:
It would be cleaner at the callsite (again) to have arguments be variadic 

`public init(_ name: Identifier, _ arguments: Expression...)`

e.g. 

`case .overflowingPlus: return .functionCall(FunctionCall("add", [lhs, rhs]))` => `case .overflowingPlus: return .functionCall(FunctionCall("add", lhs, rhs))`

Reply:
discussed elsewhere; use an array-delegating variadic overloaded constructor instead